core/store/models/ocrkey/key_bundle.go

```diff
@@ -0,0 +1,252 @@
+
+var defaultScryptParams = scryptParams{
+	N: keystore.StandardScryptN, P: keystore.StandardScryptP}
+
+var curve = secp256k1.S256()
+
+func (EncryptedKeyBundle) TableName() string {
+	return "encrypted_ocr_key_bundles"
+}
+
+// NewKeyBundle makes a new set of OCR key bundles from cryptographically secure entropy
+func NewKeyBundle() (*KeyBundle, error) {
+	reader := cryptorand.Reader
+	onChainSk, err := cryptorand.Int(reader, crypto.S256().Params().N)
+	if err != nil {
+		return nil, err
+	}
+	onChainPriv := new(onChainPrivateKey)
+	p := (*ecdsa.PrivateKey)(onChainPriv)
+	p.D = onChainSk
+	onChainPriv.PublicKey = ecdsa.PublicKey{Curve: curve}
+	p.PublicKey.X, p.PublicKey.Y = curve.ScalarBaseMult(onChainSk.Bytes())
+	_, offChainPriv, err := ed25519.GenerateKey(reader)
+	if err != nil {
+		return nil, err
+	}
+	var encryptionPriv [curve25519.ScalarSize]byte
+	_, err = reader.Read(encryptionPriv[:])
+	if err != nil {
+		return nil, err
+	}
+	k := &KeyBundle{
+		onChainSigning:     onChainPriv,
+		offChainSigning:    (*offChainPrivateKey)(&offChainPriv),
+		offChainEncryption: &encryptionPriv,
+	}
+	marshalledPrivK, err := json.Marshal(k)
+	if err != nil {
+		return nil, err
+	}
+	byteID := sha256.Sum256(marshalledPrivK)
+	k.ID = hex.EncodeToString(byteID[:])
+	return k, nil
+}
+
+// SignOnChain returns an ethereum-style ECDSA secp256k1 signature on msg.
+func (pk *KeyBundle) SignOnChain(msg []byte) (signature []byte, err error) {
+	return pk.onChainSigning.Sign(msg)
+}
+
+// SignOffChain returns an EdDSA-Ed25519 signature on msg.
+func (pk *KeyBundle) SignOffChain(msg []byte) (signature []byte, err error) {
+	return pk.offChainSigning.Sign(msg)
+}
+
+// ConfigDiffieHelman returns the shared point obtained by multiplying someone's
+// public key by a secret scalar ( in this case, the offChainEncryption key.)
+func (pk *KeyBundle) ConfigDiffieHelman(base *[curve25519.PointSize]byte) (
+	sharedPoint *[curve25519.PointSize]byte, err error,
+) {
+	p, err := curve25519.X25519(pk.offChainEncryption[:], base[:])
+	if err != nil {
+		return nil, err
+	}
+	sharedPoint = new([ed25519.PublicKeySize]byte)
+	copy(sharedPoint[:], p)
+	return sharedPoint, nil
+}
+
+// PublicKeyAddressOnChain returns public component of the keypair used in
+// SignOnChain
+func (pk *KeyBundle) PublicKeyAddressOnChain() OnChainSigningAddress {
+	return pk.onChainSigning.Address()
+}
+
+// PublicKeyOffChain returns the pbulic component of the keypair used in SignOffChain
+func (pk *KeyBundle) PublicKeyOffChain() OffChainPublicKey {
+	return OffChainPublicKey(pk.offChainSigning.PublicKey())
+}
+
+// PublicKeyConfig returns the public component of the keypair used in ConfigKeyShare
+func (pk *KeyBundle) PublicKeyConfig() [curve25519.PointSize]byte {
+	rv, err := curve25519.X25519(pk.offChainEncryption[:], curve25519.Basepoint)
+	if err != nil {
+		log.Println("failure while computing public key: " + err.Error())
+	}
+	var rvFixed [curve25519.PointSize]byte
+	copy(rvFixed[:], rv)
```

Comment:
Probably unnecessary, but it's a good idea to check that the length value returned by `copy` matches `curve25519.PointSize`.

Reply:
yeah as written this shouldn't be able to be anything but. Since this is a "getter" method I'd prefer to avoid returning an error obj, especially since it shouldn't ever error.

Reply:
Fair enough.